src/cpp/rtps/flowcontrol/FlowControllerImpl.hpp

```diff
@@ -0,0 +1,1411 @@
+        // its writer's mutex.
+        // Problematic states:
+        // - Being added: change both pointers from nullptr to a pointer values.
+        // - Being removed: change both pointer from pointer values to nullptr.
+        if (nullptr != change->writer_info.previous ||
+                nullptr != change->writer_info.next)
+        {
+            ++async_mode.writers_interested_in_remove;
+            std::unique_lock<std::mutex> lock(mutex_);
+            std::unique_lock<std::mutex> interested_lock(async_mode.changes_interested_mutex);
+
+            // When blocked, both pointer are different than nullptr or equal.
+            assert((nullptr != change->writer_info.previous &&
+                    nullptr != change->writer_info.next) ||
+                    (nullptr == change->writer_info.previous &&
+                    nullptr == change->writer_info.next));
+            if (nullptr != change->writer_info.previous &&
+                    nullptr != change->writer_info.next)
+            {
+
+                // Try to join previous node and next node.
+                change->writer_info.previous->writer_info.next = change->writer_info.next;
+                change->writer_info.next->writer_info.previous = change->writer_info.previous;
+                change->writer_info.previous = nullptr;
+                change->writer_info.next = nullptr;
+            }
+            --async_mode.writers_interested_in_remove;
+        }
+    }
+
+    /*! This function is used when PublishMode = FlowControllerPureSyncPublishMode.
+     *  In this case there is no async mechanism.
+     */
+    template<typename PubMode = PublishMode>
+    typename std::enable_if<std::is_same<FlowControllerPureSyncPublishMode, PubMode>::value, void>::type
+    remove_change_impl(
+            fastrtps::rtps::CacheChange_t*) const
+    {
+        // Do nothing. Fail.
+    }
+
+    /*!
+     * Function ran by asynchronous thread.
+     */
+    void run()
+    {
+        while (async_mode.running)
+        {
+            // There is writers interested in remove a sample.
+            if (0 != async_mode.writers_interested_in_remove)
+            {
+                continue;
+            }
+
+            std::unique_lock<std::mutex> lock(mutex_);
+            fastrtps::rtps::CacheChange_t* change_to_process = nullptr;
+
+            //Check if we have to sleep.
+            {
+                std::unique_lock<std::mutex> in_lock(async_mode.changes_interested_mutex);
+                // Add interested changes into the queue.
+                sched.add_interested_changes_to_queue_nts();
+
+                while (async_mode.running &&
+                        (async_mode.force_wait() || nullptr == (change_to_process = sched.get_next_change_nts())))
+                {
+                    lock.unlock();
+                    bool ret = async_mode.wait(in_lock);
+
+                    in_lock.unlock();
+                    lock.lock();
+                    in_lock.lock();
```

Comment:
Please document why main lock should be released.

Reply:
Done